.github/workflows/build.yaml

```diff
@@ -35,6 +35,41 @@ jobs:
         with:
           node-version: "16"
 
+      - name: Fetch dependencies from cache
+        id: cache-node-modules
+        uses: actions/cache@v3
+        with:
+          path: "node_modules"
+          key: yarn-build-${{ hashFiles('yarn.lock') }}
+          restore-keys: |
+            yarn-build-
+
+      - name: Install dependencies
```

Comment:
I think it should be okay. From the testing I've done it hasn't seemed to be a problem (though I haven't tried changing a dependency). I try that now though and see.

Reply:
https://github.com/coder/code-server/actions/runs/3093663725

Seems like it's okay as is to me. What do you think?

Reply:
Only one of them ran since the rest have no changed files so it is hard to tell if there will be a problem in cases where multiple of them will runp

Reply:
If they did all run though it looks like it would take about an extra two minutes (30s per cache restore and install) and idk about the cache upload, some of them might fail or it time out or maybe overwrite each other.

Reply:
Hmm...talking out loud for a second, these are the jobs running in parallel:
- "Format with Prettier"
- "Lint helm chart"
- "Lint shell files"
- "Lint TS files"
- "Check TypeScript types"

3 rely on `node-modules` - prettier and two TS jobs. 30s per cache restore x 3 = 1.5 mins worst-case scenario if the change affects all 3. Example of that would be change to `*.ts` file _and_ a change to a `yarn.lock`. I don't think `yarn.lock` changes too often but worth testing. I'll add a test commit for this.

> The cache is scoped to the key and branch. The default branch cache is available to other branches.

Didn't realize this. Since they all use the same key, they could overwrite each other. We could wait until we have evidence that problem exists before we preemptively solve for it?

Reply:
We should test that these workflows actually run when changes are made.  It sounds like they might not be?

Reply:
> We could wait until we have evidence that problem exists before we preemptively solve for it?

IMO this is something we can easily test now before merging.

Reply:
Dude, you the man! Thanks for checking that conflict. Great news 🎉

Okay let me take these action items and address them so we can move this forward:
- [ ] rename `scripts.yml` to `scripts.yaml`
- [ ] move lint scripts to that workflow
- [ ] use `apt` so we don't need node
- [ ] fix typecheck step based on your run
- [ ] fix lint typescript step not finding changes

Reply:
First three definitely optional, totally down to leave as-is if you wanna get this PR merged ASAP.

For the typecheck script actually I am curious for the reasoning because we would already be doing it in the build step.

Reply:
> For the typecheck script actually I am curious for the reasoning because we would already be doing it in the build step

Originally added it for two reasons:
1. because i was going to remove it from our tests and move away from `ts-jest`
2. faster feedback instead of waiting for the Build step (but probably doesn't happen much faster)

I don't think those are good enough reasons now that I'm writing this though so I'll remove it

Reply:
Also I was editing my comment like crazy, let me know if you have any thoughts on the cancel action portion.

Reply:
>  because i was going to remove it from our tests and move away from ts-jest
>  faster feedback instead of waiting for the Build step (but probably doesn't happen much faster)

Very fair points, we should definitely revisit a separate typecheck step when we investigate alternate build tools.

Reply:
> And it could be nice to verify if the build at least works even if you have, for example, a helm error (to avoid the pattern where you only see one error from a job at a time and have to keep waiting to see each error rather than see them all at once).

I see what you mean. Wouldn't you end up having to rebuild regardless though after you push a fix for the helm linting error? My thinking was at least you'd see CI in a "done" state faster and be able to fix it but maybe not. I could go both ways.

Reply:
Also if you think about it...the "cancel workflow" kinda mimics what we had before when it was called "Prebuild". "Build" only ran if "Prebuild" passed so it was essentially the same flow (not saying we have to keep it but wanted to mention).

Reply:
I can't reproduce your TS error locally 🤔


```shell
❯ yarn tsc -p tsconfig.json --noEmit --skipLibCheck
yarn run v1.22.18
$ /Users/jp/dev/oss/code-server/node_modules/.bin/tsc -p tsconfig.json --noEmit --skipLibCheck
✨  Done in 1.77s.
```

Reply:
oh wow...so i was looking at our current [`lint.sh`](https://github.com/coder/code-server/blob/main/ci/dev/lint.sh) and it has:
```shell
tsc --noEmit --skipLibCheck
```

_that's_ why I added the typecheck step 🤦🏼‍♂️

Reply:
@code-asher I think it's ready for another review.

Reply:
> the "cancel workflow" kinda mimics what we had before when it was called "Prebuild"

True!  I think the main difference is that the cancel action seems to fail on forks and prevent cache uploads.

>  Wouldn't you end up having to rebuild regardless though after you push a fix for the helm linting error?

Yup, but my thinking is that once a check shows a red X you have two choices:

1. Wait for the rest of CI to see if anything else went wrong
2. Immediately fix and push the lint error

Having the option of the first is nice because depending on the timing of the parallel steps you could run into something like:

Ah I have a Helm error.  The workflow was canceled so no idea if there are any other errors.  I will push up the Helm fix, wait for the build again.  Ah I have a Typescript lint error, I will fix that.  Ah this time I have a build error.

Canceling CI automatically forces the second option.  And without the automatic cancel you can still take the second option if you want and the cancel will happen naturally since we have that cancel thing when you re-push.

> I can't reproduce your TS error locally

Weird...maybe my branch is just borked somehow.  I can try again.

> that's why I added the typecheck step

Ohhhhhhhhhh dude I had no idea that was even in there!  That explains it.

Reply:
> Canceling CI automatically forces the second option. And without the automatic cancel you can still take the second option if you want and the cancel will happen naturally since we have that cancel thing when you re-push.

Ah, that's a good point! Okay I'm on board. Let's remove it.

Reply:
Should be good now 👍🏼